### PR TITLE
Implement a Calico IPAM driver

### PR DESCRIPTION
**Pull request by robbrockbank**

### CODE REVIEW

**Comment by tomdee** [comment]
Circle failing and conflicts. Can you rebase maybe with a better title and description :smile:

File: libnetwork/driver_plugin.py [hunk 1]
```diff
@@ -12,27 +12,53 @@
...
-from netaddr import IPNetwork
+from netaddr import IPAddress, IPNetwork
 from pycalico.datastore_datatypes import IF_PREFIX, Endpoint, IPPool
 from pycalico import netns
+from pycalico.util import get_hostname
 
 from datastore_libnetwork import LibnetworkDatastoreClient
 
+
+# Fixed MAC address used in the container
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: You've added a load of comments, but IMHO they're pretty pointless. If you're going to add them I think you should take them a bit further. e.g. I can see that FIXED_MAC is a fixed mac address - can your comment say why we're using a fixed mac address and why we've chosen that value?

File: libnetwork/driver_plugin.py [hunk 2]
```diff
@@ -12,27 +12,53 @@
...
 FIXED_MAC = "EE:EE:EE:EE:EE:EE"
+
+# Orchestrator and container IDs used in our endpoint identification. These
+# are fixed for libnetwork.
 CONTAINER_NAME = "libnetwork"
 ORCHESTRATOR_ID = "libnetwork"
 
+# Fixed global and local address spaces.
+LOCAL_CIDR_ADDR_SP = "CalicoLocalPool"
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: Do you charge by the character? What's with the abbreviated names?

File: libnetwork/driver_plugin.py [hunk 3]
```diff
@@ -12,27 +12,53 @@
...
+GLOBAL_CIDR_ADDR_SP = "CalicoGlobalPool"
+
+# Fixed pool IDs.
+POOL_ID_V4 = "CalicoPoolIPv4"
+POOL_ID_V6 = "CalicoPoolIPv6"
+
+# Fixed pool CIDRs.
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: I think these really need more explanation - why these values?

File: libnetwork/driver_plugin.py [hunk 4]
```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
...
     return jsonify(json_response)
 
+# <-- IPAM plugin API -->
+
+@app.route('/IpamDriver.GetDefaultAddressSpaces', methods=['POST'])
+def get_default_address_spaces():
+    json_data = request.get_json(force=True)
+    app.logger.debug("GetDefaultAddressSpace JSON=%s", json_data)
+    json_response = {
+        "LocalDefaultAddressSpace": LOCAL_CIDR_ADDR_SP,
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: I'm not the biggest fan of having constants defined somewhere else when 
- they're only used once
- it's only a small program.

It's just bloating the code and making the reader jump around unnecessarily.

File: libnetwork/driver_plugin.py [hunk 5]
```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
...
+    json_data = request.get_json(force=True)
+    app.logger.debug("RequestPool JSON=%s", json_data)
+    address_space = json_data["AddressSpace"]
+    pool = json_data["Pool"]
+    sub_pool = json_data.get("SubPool")
+    options = json_data["Options"]
+    v6 = json_data["V6"]
+
+    # Calico IPAM does not allow you to request Pool or SubPool.
+    assert not pool and not sub_pool, \
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: Is assert the right thing to use here? The API defines an error response which we try to confirm to using make_json_error(). Can you confirm that using assert gives the right user experience. This is easy to check manually which you should do in addition to adding a UT.

File: libnetwork/driver_plugin.py [hunk 6]
```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
...
+        pool_cidr = POOL_CIDR_STR_V6
+        gway_cidr = GWAY_CIDR_STR_V6
+    else:
+        pool_id = POOL_ID_V4
+        pool_cidr = POOL_CIDR_STR_V4
+        gway_cidr = GWAY_CIDR_STR_V4
+
+    # The meta data includes a dummy gateway address.  This prevents libnetwork
+    # from requesting a gateway address from the pool since for a Calico
+    # network we route to the host.
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: ...which means the route/gateway is set up per container (in join()) rather than per network.

File: libnetwork/driver_plugin.py [hunk 7]
```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
...
+        "network'.  Calico IP Pools should be configured first and IP " \
+        "assignment is from those pre-configured pools."
+
+    # We use static pool ID and CIDR - these are arbitrary since we do not
+    # actually create a pool here.  It is assumed the user has configured
+    # all of the required Calico pools.
+    if v6:
+        pool_id = POOL_ID_V6
+        pool_cidr = POOL_CIDR_STR_V6
+        gway_cidr = GWAY_CIDR_STR_V6
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: Can't you just say gateway_cidr?

File: libnetwork/driver_plugin.py [hunk 8]
```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
...
+
+        # Auto assign an IP based on whether the IPv4 or IPv6 pool was selected.
+        # We auto-assign from all available pools with affinity based on our host.
+        ips_v4, ips_v6 = client.auto_assign_ips(num_v4, num_v6, None, None,
+                                                pool=(None, None),
+                                                hostname=hostname)
+    else:
+        # A specific IP has been requested, so assign from our pools.  We
+        # do not require an affine block in this case.
+        client.assign_ip(IPAddress(address), None, {}, hostname=hostname)
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: What happens in the address is outside a Calico pool?

File: libnetwork/driver_plugin.py [hunk 9]
```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
...
+
+    ips = ips_v4 + ips_v6
+    assert ips, "No IP addresses available in the configured pools"
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: Again can you confirm that assert is the right thing to use?

Comment (tomdee) [inline libnetwork/driver_plugin.py]: And might it be worth asserting that the list contains exactly one element?

File: libnetwork/driver_plugin.py (hunk 5 above)
Comment (robbrockbank) [inline libnetwork/driver_plugin.py]: In the case I tried - assert did have the right user experience, but I would rather do something different - was going to ask if there was a preferred exception to throw that you knew about.  If you don't I'll dig - assert is deffo not the correct final solution.

File: libnetwork/driver_plugin.py [hunk 10]
```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
...
+
+@app.route('/IpamDriver.ReleaseAddress', methods=['POST'])
+def release_address():
+    json_data = request.get_json(force=True)
+    app.logger.debug("ReleaseAddress JSON=%s", json_data)
+    pool_id = json_data["PoolID"]
+    address = json_data["Address"]
+
+    # Unassign the address.
+    client.release_ips({IPAddress(address)})
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: Do we need to retry if the release fails?

File: libnetwork/driver_plugin.py [hunk 11]
```diff
@@ -98,18 +242,28 @@ def create_network():
...
         ip_data = json_data["IPv%sData" % version]
-        if ip_data:
-            client.add_ip_pool(version, IPPool(ip_data[0]['Pool'],
-                                               ipip=ipip,
-                                               masquerade=masquerade))
+        if not ip_data:
+            # No IP data for this IP version, so skip.
+            app.logger.info("No IPv%s data", version)
+            continue
+        gway = IPNetwork(ip_data[0]['Gateway']).ip
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: gateway

File: libnetwork/driver_plugin.py (hunk 2 above)
Comment (robbrockbank) [inline libnetwork/driver_plugin.py]: Old habits die hard.....

File: libnetwork/driver_plugin.py [hunk 12]
```diff
@@ -192,12 +358,23 @@ def join():
         remove_veth(host_interface_name)
         raise e
 
+    # TODO Optionally use the docker bridge, don't specify gateway
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: This needs to be resolved before you can merge.

We shouldn't provide the staticroutes if the gateway is blank etc...

File: tests/st/libnetwork/test_mainline_multi_host.py [hunk 13]
```diff
@@ -24,7 +24,31 @@
 
 class MultiHostMainline(TestBase):
 
-    def test_multi_host(self):
+    def test_multi_host_default_ipam(self):
+        """
+        Run a mainline multi-host test.
+
+        Because multihost tests are slow to setup, this tests most mainline
```
Comment (tomdee) [inline tests/st/libnetwork/test_mainline_multi_host.py]: Comment seems odd to repeat twice.

File: libnetwork/driver_plugin.py (hunk 12 above)
Comment (robbrockbank) [inline libnetwork/driver_plugin.py]: Yup - this tied in with what you were doing - wanted to touch base with you on this first.

File: libnetwork/driver_plugin.py [hunk 14]
```diff
@@ -70,13 +96,131 @@ def make_json_error(ex):
...
-    json_response = {"Implements": ["NetworkDriver"]}
+    json_response = {"Implements": ["NetworkDriver", "IpamDriver"]}
     app.logger.debug("Activate response JSON=%s", json_response)
     return jsonify(json_response)
 
+# <-- IPAM plugin API -->
+
+@app.route('/IpamDriver.GetDefaultAddressSpaces', methods=['POST'])
+def get_default_address_spaces():
+    json_data = request.get_json(force=True)
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: The force comment from line 234 should be moved up here.

**Comment by tomdee** [comment]
Finished a first pass. Looks good. A few nitpicks and the return JSON of the join to sort out.

Were you planning to write more STs? You've already said you're going to do UTs. Can you update the README.md too.

File: libnetwork/driver_plugin.py (hunk 8 above)
Comment (robbrockbank) [inline libnetwork/driver_plugin.py]: Yup I need to handle that, and also the case when the IP is already assigned.

File: libnetwork/driver_plugin.py (hunk 10 above)
Comment (robbrockbank) [inline libnetwork/driver_plugin.py]: The release_ips already performs a number of retries.  In the event this fails it is due to continual compare and swap errors in which case all bests are off anyways.  So I don't think we need to explicitly retry here.

File: libnetwork/driver_plugin.py [hunk 15]
```diff
@@ -12,27 +12,58 @@
...
+
+# Fix pool and gateway CIDRs.  As per comment above, Calico IPAM does not allow
+# assignment from a specific pool, so we choose a dummy value that will not be
+# used in practise.  A 0/0 value is used for both IPv4 and IPv6.  This value is
+# also used by the Network Driver to indicate that the Calico IPAM driver was
+# used rather than the default libnetwork IPAM driver - this is useful because
+# Calico Network Driver behavior depends on whether our IPAM driver was used or
+# not.
+#
+# Implementation note:  Boolean operatorion is used to determine a 0/0 CIDR, so
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: typo

File: libnetwork/driver_plugin.py [hunk 16]
```diff
@@ -70,13 +101,154 @@ def make_json_error(ex):
...
+
+
+@app.route('/IpamDriver.ReleaseAddress', methods=['POST'])
+def release_address():
+    json_data = request.get_json(force=True)
+    app.logger.debug("ReleaseAddress JSON=%s", json_data)
+    address = json_data["Address"]
+
+    # Unassign the address.
+    client.release_ips({IPAddress(address)})
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: Can you expand the comment to reassure the reader that it's OK if the release_ips fails. What happens if the release_ips throws an exception and will it generate a nice error for the user that's removing their container.

File: libnetwork/driver_plugin.py [hunk 17]
```diff
@@ -199,14 +383,54 @@ def join():
...
-        "Gateway": "",  # Leave gateway empty to trigger auto-gateway behaviour
+        }
     }
 
+    # Extract relevant data from the Network data.
+    network_data = get_network_data(network_id)
+    gateway_ip4, _ = get_gateway_pool_from_network_data(network_data, 4)
+    gateway_ip6, _ = get_gateway_pool_from_network_data(network_data, 6)
+
+    #if (gateway_ip4 and is_using_calico_ipam(gateway_ip4)) or \
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: TODO - Sort this out.

File: libnetwork/driver_plugin.py [hunk 18]
```diff
@@ -306,3 +531,63 @@ def remove_veth(name):
         netns.remove_veth(name)
     except CalledProcessError:
         app.logger.warn("Failed to delete veth %s", name)
+
+
+def get_network_data(network_id):
+    """
+    Return the network data, and raise an exception if the network does not
+    exists.
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: typo

File: libnetwork/driver_plugin.py [hunk 19]
```diff
@@ -306,3 +531,63 @@ def remove_veth(name):
         netns.remove_veth(name)
     except CalledProcessError:
         app.logger.warn("Failed to delete veth %s", name)
+
+
+def get_network_data(network_id):
+    """
+    Return the network data, and raise an exception if the network does not
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: Might be nicer to say what "network data" actually is - i.e. the JSON that's passed in on the CreateNetwork request.

File: libnetwork/driver_plugin.py [hunk 20]
```diff
@@ -306,3 +531,63 @@ def remove_veth(name):
...
+    assert version in [4,6]
+
+    ip_data = network_data.get("IPv%sData" % version)
+    if not ip_data:
+        # No IP data for this IP version, so skip.
+        app.logger.info("No IPv%s data", version)
+        return None, None
+
+    if len(ip_data) > 1:
+        error_message = "Unsupported: mutliple Gateways defined for " \
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: type

File: libnetwork/driver_plugin.py [hunk 21]
```diff
@@ -306,3 +531,63 @@ def remove_veth(name):
...
+def is_using_calico_ipam(gateway_cidr):
+    """
+    Determine if the gateway CIDR indicates that we are using Calico IPAM
+    driver for IP assignment.
+    :param gateway: The gateway CIDR.  This should not be None.
+    :return: True, if using Calico IPAM driver.
+    """
+    # A 0 gateway IP indicates Calico IPAM is being used.
+    assert isinstance(gateway_cidr, IPNetwork)
+    return (not gateway_cidr.ip) and not (gateway_cidr.prefixlen)
```
Comment (tomdee) [inline libnetwork/driver_plugin.py]: What's the reasoning behind doing this boolean comparison? Can't you just do an equality check on an IPNetwork object directly?

**Comment by tomdee** [comment]
@robbrockbank Can you summarize the issues that are stopping us from merging this is the PR description.

**Comment by robbrockbank** [comment]
Current issues preventing us from merging are:
-  temporary pdb breakpoint in the mainline multi host test
-  multi host test is failing because containers created with calico ipam and default ipam are not correctly isolated.
-  Container create using default IPAM can ping container created with calico ipam _until_ felix does a resync which updates the ip tables and correct isolation results.  Need to have a think about ip tables interactions between felix and libnetwork.
-  Need to handle separation of IP Pools allocated when using default IPAM.  The IP pools used by default IPAM and calico IPAM should not be mixed since we do not have control over IP allocations in the pools used by default IPAM.

**Comment by tomdee** [comment]
How about handling that last bullet point as a separate PR? You can start working on that now, and I can take on resolving the other stuff (when I've finished doing a calico-docker release)

**Comment by robbrockbank** [comment]
@tomdee  Yes that makes sense - in the meantime mixing calico and default ipam is a no-no.

File: README.md [hunk 22]
```diff
@@ -50,13 +54,31 @@ DiscoverNew    | 0     | 0     | 0      | None
...
+DeleteNetwork  | 1     | 0     | 3 (4 if IPv4 and IPv6)     | Delete profile, pool and stored CreateNetwork JSON
+DeleteEndpoint | 0     | 0     | 1      | Delete endpoint
+Leave          | 0     | 0     | 0      | None
+
+
+Datastore interactions using Calico IPAM:
+
+Operation      | Reads | Writes| Deletes| Notes
+---------------|-------|-------|--------|------
+RequestPool    | 0     | 0     | 0      | None
```
Comment (tomdee) [inline README.md]: Can you make the ordering of operations the same on both tables?

File: README.md [hunk 23]
```diff
@@ -50,13 +54,31 @@ DiscoverNew    | 0     | 0     | 0      | None
...
+ReleasePool    | 0     | 0     | 0      | None
+RequestAddress | >=2   | >=1   | >=0    | May have multiple reads/writes/deletes depending on contention (see libcalico IPAM)
+ReleaseAddress | >=1   | >=1   | >=0    | May have multiple reads/writes/deletes depending on contention (see libcalico IPAM)
+DiscoverNew    | 0     | 0     | 0      | None
+DiscoverDelete | 0     | 0     | 0      | None
+CreateNetwork  | 0     | 3     | 0      | 2 for creating profile (tags and rules), and 1 to store the request JSON
+CreateEndpoint | 3     | 1     | 0      | Read CreateNetwork JSON and IPv4/IPv6 next hops, and write Endpoint
```
Comment (tomdee) [inline README.md]: Isn't getting the nexthops only a single read?

File: README.md [hunk 24]
```diff
@@ -50,13 +54,31 @@ DiscoverNew    | 0     | 0     | 0      | None
...
+Join           | 0     | 0     | 0      | Read CreateNetwork JSON and Endpoint
```
Comment (tomdee) [inline README.md]: Why is Reads zero when the comment says otherwise?

File: tests/st/libnetwork/test_mainline_multi_host.py [hunk 25]
```diff
@@ -46,46 +51,95 @@ def test_multi_host(self):
...
-            workload_host1.execute("ping -c 1 -W 1 workload2")
-            workload_host2.execute("ping -c 1 -W 1 workload1")
+            assert_number_endpoints(host1, 4)
+            assert_number_endpoints(host2, 2)
+
+            # Assert that workloads can communicate with each other on network
+            # 1, and not those on network 2.  Ping using IP for all workloads,
+            # and by hostname for workloads on the same network (note that
+            # a workloads own hostname does not work).
+            workload_h1n1_1.execute("cat /etc/hosts")
```
Comment (tomdee) [inline tests/st/libnetwork/test_mainline_multi_host.py]: Should these cat /etc/hosts lines be checked in?

File: tests/st/libnetwork/test_mainline_multi_host.py [hunk 26]
```diff
@@ -46,46 +51,95 @@ def test_multi_host(self):
...
+            workload_h2n1_1.execute("cat /etc/hosts")
+            self.assert_connectivity(pass_list=[workload_h1n1_1,
+                                                workload_h1n1_2,
+                                                workload_h2n1_1])
+            # TODO: docker_gwbridge iptable FORWARD rule takes precedence over
```
Comment (tomdee) [inline tests/st/libnetwork/test_mainline_multi_host.py]: Are you raising an issue with libnetwork for this so you can raise the issue here?

**Comment by tomdee** [comment]
OK, a few more small comments but otherwise LGTM.

Do you think it's worth mentioning some of the restrictions in the README.md? e.g.
- don't mix the two types of IPAM
- Can only use a single network per container with Calico-ipam (but can manually assign multiple profiles)

File: README.md (hunk 23 above)
Comment (robbrockbank) [inline README.md]: Nope, 2.

File: tests/st/libnetwork/test_mainline_multi_host.py (hunk 25 above)
Comment (robbrockbank) [inline tests/st/libnetwork/test_mainline_multi_host.py]: I hadn't intended to take them out - thought it provides some useful diags incase host name stuff ever goes wrong.  But I can take it out if you'd prefer.

File: tests/st/libnetwork/test_mainline_multi_host.py (hunk 26 above)
Comment (robbrockbank) [inline tests/st/libnetwork/test_mainline_multi_host.py]: i haven't but i will.

File: README.md (hunk 24 above)
Comment (robbrockbank) [inline README.md]: yup, should be 2 as per the text
